src/main/kotlin/com/mdrobnak/lalrpop/inspections/WrongInferredTypeInspection.kt

```diff
@@ -0,0 +1,36 @@
+package com.mdrobnak.lalrpop.inspections
+
+import com.intellij.codeInspection.LocalInspectionTool
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.psi.PsiElementVisitor
+import com.mdrobnak.lalrpop.psi.LpNonterminal
+import com.mdrobnak.lalrpop.psi.LpVisitor
+
+/**
+ * Reports 2 kinds of problems on a nonterminal
+ * 1. The inferred type is different from the type declared explicitly on the nonterminal
+ * 2. There are 2 different alternatives, without action code, where the inferred types are different
+ */
+object WrongInferredTypeInspection : LocalInspectionTool() {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
+        return object : LpVisitor() {
+            override fun visitNonterminal(nonterminal: LpNonterminal) {
+                val explicitType = nonterminal.typeRef?.resolveType(listOf())
+
+                var seenType = explicitType
+
+                nonterminal.alternatives.alternativeList.filter { it.action == null }.forEach { alternative ->
+                    val alternativeType = alternative.resolveType(listOf())
+
+                    if (seenType != null && seenType != alternativeType) { // TODO: maybe get the rust plugin to do this comparison because it may be very error-prone by just comparing strings
```

Comment:
I have no idea how to do this. Would appreciate any help.

Reply:
One approach would be to create a PSI file with Rust code so the Rust plugin parses and type checks it. Then get the types and compare them.

Example of Rust template to use:
```rust
// Imports here

type TypeToCompare1 = /* Place type string here */;
type TypeToCompare2 = /* Place type string here */;
```
Then we can compare the types of `TypeToCompare1` and `TypeToCompare2`.

Here's some APIs I found in intellij-rust that might be helpful:
[`RsTypeAlias::declaredType`](https://github.com/intellij-rust/intellij-rust/blob/31f6caf7471a189be0aee0c00cdd6a7f4107eab0/src/main/kotlin/org/rust/lang/core/psi/ext/RsTypeAlias.kt#L50) returns a `Ty` which might represent the aliased type.
[`TypeInference::combineTypesNoVars`](https://github.com/intellij-rust/intellij-rust/blob/17880fcf784d39f47b0178184a0f10e393da69c6/src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt#L439) might be useful for comparing types, but I haven't found a general method to do so. We might want to make an issue on intellij-rust for help if we can't find a general method ourselves.

Reply:
Great! Will play around with them tomorrow and I'll see if I can get something to work

Reply:
I think I mostly got it to work, only that I decided to import it to gluon's grammar and it has 2 nonterminals defined like:
```
Any: () = {
    <"identifier">,
    <"operator">,
    <"string literal">,
    <"char literal">,
    <"int literal">,
    <"byte literal">,
    <"float literal">,

    <",">,
    <"=">,

    "(" AttributeContents ")",
    "[" AttributeContents "]",
    "{" AttributeContents "}",
};

AttributeContents: () = {
    Any*
};
```

Here, the inspection fails and reports on every alternative that the type is wrong, so maybe `()` should be special cased?

Reply:
Yep, looks like lalrpop special-cases the unit type if there's no action code:
https://github.com/Mcat12/intellij-lalrpop/pull/22#discussion_r536817454